### ISSUE TITLE
[CDC + Before Image] Before image is not as expected after some time, CDC stops

### ISSUE THREAD

**Issue by shamanthchandra-yb** [yugabyte/yugabyte-db#15309]
Jira Link: [DB-4461](https://yugabyte.atlassian.net/browse/DB-4461)
### Description

Two runs to prove it:

http://stress.dev.yugabyte.com/stress_test/cfaa4811-9d4b-4e22-9b1e-c23161d0cd23

http://stress.dev.yugabyte.com/stress_test/ce9323ed-521c-435c-8136-512ddb4fa9ca

![Screenshot 2022-12-13 at 2 20 22 PM](https://user-images.githubusercontent.com/109518123/207270184-4aed491e-922f-4da8-8fe7-576671aa109a.png)

**Comment by sureshdash2022-yb** [comment]
Hi @shamanthchandra-yb 
To retain the before image for the table  CDC uses `cdcsdk_safe_time`. CDC updates at the tablet level so that compaction can be restricted for the rows created or modified after the `cdcsdk_safe_time` point. 

To read the before image for the UPDATE/DELETE records, CDC uses the transaction commit timestamp as read time to get the before image from the actual database. 

In this failure scenario, we have observed that we are running large transactions with UPDATE/DELETE operations, whereas CDC read changes from intentsDB, which it can read at time  FLAGS_cdc_max_stream_intent_records(1680) intent records, so for the current batch CDC use transaction commit timestamp as read timestamp to get the before image for the changed records, but the pending intents, CDC resume from the from_op_id(which is the same op_id of the current batch). As part of the next `Getchanges` call, it should get the before image for the pending intents, using the transaction committed timestamp of previously processed intents. But if the last record in the WAL log is the `UPADTE_TRASACTION_OP` record, for the above scenario as part of pending intents reading CDC is getting 0 read_ops records from the WAL, which is causing the before image to be missing.


Apart from this, we have identified a few more gaps in the code, here is a list of issues:- 
For a larger transaction where intents are more than `FLAGS_cdc_max_stream_intent_records`, CDC should set the `cdcsdk_safe_time` as the last processed intent record’s commit timestamp, which is wrongly updated as `leader_safe_time` which will cause before image compaction.

 For the pending intents reading CDC should use the transaction commit timestamp, not the record hybrid_time.
